### PR TITLE
[SPARK-38257][BUILD] Upgrade `rockdbjni` to 7.0.3

### PR DESCRIPTION
**Pull request by LuciferYang**
### What changes were proposed in this pull request?
The main change of this pr is upgrade `rockdbjni` from 6.20.3 to 7.0.3.

### Why are the changes needed?
`rockdbjni` 7.x is a Java 8 compiled version, while `rockdbjni` 6.x  is still a Java 7 compiled version, At the same time `rockdbjni` 7.x supports MacOS universal binary (ARM+x86):

- https://github.com/facebook/rocksdb/releases/tag/v6.22.1
- https://github.com/facebook/rocksdb/releases/tag/v6.23.3
- https://github.com/facebook/rocksdb/releases/tag/v6.24.2
- https://github.com/facebook/rocksdb/releases/tag/v6.25.1
- https://github.com/facebook/rocksdb/releases/tag/v6.25.3
- https://github.com/facebook/rocksdb/releases/tag/v6.26.0
- https://github.com/facebook/rocksdb/releases/tag/v6.26.1
- https://github.com/facebook/rocksdb/releases/tag/v6.27.3
- https://github.com/facebook/rocksdb/releases/tag/v6.28.2
- https://github.com/facebook/rocksdb/releases/tag/v6.29.3
- https://github.com/facebook/rocksdb/releases/tag/v6.29.4
- https://github.com/facebook/rocksdb/releases/tag/v7.0.1
- https://github.com/facebook/rocksdb/releases/tag/v7.0.2
- https://github.com/facebook/rocksdb/releases/tag/v7.0.3

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?

- Pass GA
- Manual performance test using `org.apache.spark.util.kvstore.RocksDBBenchmark`

environment

```
OS: CentOS release 6.3 (Final)
Kernel: 4.14
CPU:  Intel(R) Xeon(R) Gold @ 2.60GHz
``` 

**Before**

6.20.3
```
[INFO] Running org.apache.spark.util.kvstore.RocksDBBenchmark
                                        	count	mean	min	max	95th
dbClose                                 	4	0.275	0.174	0.484	0.484
dbCreation                              	4	74.612	3.318	289.818	289.818
naturalIndexCreateIterator              	1024	0.005	0.002	1.470	0.007
naturalIndexDescendingCreateIterator    	1024	0.004	0.003	0.069	0.006
naturalIndexDescendingIteration         	1024	0.006	0.004	0.039	0.009
naturalIndexIteration                   	1024	0.006	0.004	0.052	0.010
randomDeleteIndexed                     	1024	0.025	0.019	0.228	0.034
randomDeletesNoIndex                    	1024	0.015	0.012	0.037	0.017
randomUpdatesIndexed                    	1024	0.091	0.039	30.743	0.101
randomUpdatesNoIndex                    	1024	0.063	0.059	0.492	0.068
randomWritesIndexed                     	1024	0.117	0.033	52.323	0.121
randomWritesNoIndex                     	1024	0.068	0.022	1.346	0.074
refIndexCreateIterator                  	1024	0.004	0.004	0.017	0.006
refIndexDescendingCreateIterator        	1024	0.003	0.003	0.027	0.004
refIndexDescendingIteration             	1024	0.006	0.005	0.047	0.008
refIndexIteration                       	1024	0.008	0.005	0.279	0.011
sequentialDeleteIndexed                 	1024	0.020	0.017	0.125	0.025
sequentialDeleteNoIndex                 	1024	0.014	0.012	0.040	0.017
sequentialUpdatesIndexed                	1024	0.077	0.065	1.297	0.101
sequentialUpdatesNoIndex                	1024	0.069	0.049	0.722	0.084
sequentialWritesIndexed                 	1024	0.076	0.059	2.006	0.087
sequentialWritesNoIndex                 	1024	0.063	0.025	2.529	0.074
```

**After** 

7.0.3

```
[INFO] Running org.apache.spark.util.kvstore.RocksDBBenchmark
                                        	count	mean	min	max	95th
dbClose                                 	4	0.332	0.235	0.557	0.557
dbCreation                              	4	75.354	3.301	290.771	290.771
naturalIndexCreateIterator              	1024	0.005	0.002	1.418	0.007
naturalIndexDescendingCreateIterator    	1024	0.006	0.006	0.068	0.008
naturalIndexDescendingIteration         	1024	0.006	0.004	0.261	0.007
naturalIndexIteration                   	1024	0.006	0.004	0.061	0.009
randomDeleteIndexed                     	1024	0.026	0.020	0.231	0.035
randomDeletesNoIndex                    	1024	0.014	0.012	0.040	0.017
randomUpdatesIndexed                    	1024	0.081	0.032	31.135	0.085
randomUpdatesNoIndex                    	1024	0.037	0.033	0.528	0.042
randomWritesIndexed                     	1024	0.119	0.034	54.771	0.122
randomWritesNoIndex                     	1024	0.042	0.036	1.448	0.048
refIndexCreateIterator                  	1024	0.005	0.004	0.020	0.006
refIndexDescendingCreateIterator        	1024	0.003	0.003	0.029	0.005
refIndexDescendingIteration             	1024	0.006	0.005	0.046	0.008
refIndexIteration                       	1024	0.007	0.005	0.072	0.009
sequentialDeleteIndexed                 	1024	0.021	0.017	0.100	0.026
sequentialDeleteNoIndex                 	1024	0.014	0.012	0.039	0.017
sequentialUpdatesIndexed                	1024	0.047	0.040	0.747	0.058
sequentialUpdatesNoIndex                	1024	0.040	0.029	0.730	0.049
sequentialWritesIndexed                 	1024	0.050	0.044	1.842	0.057
sequentialWritesNoIndex                 	1024	0.041	0.031	2.213	0.044
```

Additional 7.0.3 on Apple silicon/M1

```
OS: Big Sur 11.4
Kernel: 20.5.0
CPU:  Apple M1
``` 

```
[INFO] Running org.apache.spark.util.kvstore.RocksDBBenchmark
                                        	count	mean	min	max	95th
dbClose                                 	4	0.445	0.400	0.500	0.500
dbCreation                              	4	359.993	169.192	886.616	886.616
naturalIndexCreateIterator              	1024	0.003	0.001	0.902	0.003
naturalIndexDescendingCreateIterator    	1024	0.003	0.003	0.052	0.003
naturalIndexDescendingIteration         	1024	0.004	0.003	0.085	0.010
naturalIndexIteration                   	1024	0.004	0.003	0.175	0.005
randomDeleteIndexed                     	1024	0.020	0.012	0.335	0.053
randomDeletesNoIndex                    	1024	0.011	0.009	0.034	0.013
randomUpdatesIndexed                    	1024	0.043	0.020	14.436	0.040
randomUpdatesNoIndex                    	1024	0.022	0.015	0.480	0.037
randomWritesIndexed                     	1024	0.091	0.022	50.383	0.088
randomWritesNoIndex                     	1024	0.042	0.021	7.483	0.057
refIndexCreateIterator                  	1024	0.002	0.002	0.317	0.002
refIndexDescendingCreateIterator        	1024	0.003	0.001	0.624	0.002
refIndexDescendingIteration             	1024	0.009	0.003	4.596	0.010
refIndexIteration                       	1024	0.009	0.003	1.080	0.009
sequentialDeleteIndexed                 	1024	0.015	0.013	0.046	0.018
sequentialDeleteNoIndex                 	1024	0.010	0.009	0.237	0.011
sequentialUpdatesIndexed                	1024	0.026	0.020	0.622	0.032
sequentialUpdatesNoIndex                	1024	0.022	0.017	0.586	0.026
sequentialWritesIndexed                 	1024	0.048	0.027	3.646	0.063
sequentialWritesNoIndex                 	1024	0.035	0.023	2.910	0.043
```

### CODE REVIEW

**Comment by LuciferYang** [comment]
It seems that the Maven central repository has not been synchronized to the release, and it will take a few days. Set this pr to `Draft` first

**Comment by LuciferYang** [comment]
After upgrade `rockdbjni` from 6.20.2 to 6.29.3 in this pr, then we can enable all  UTs related to `rocksdb` on Apple silicon,  is it better to enable these UTs in a new pr or to complete it together in this pr? what do you think @dongjoon-hyun ?

**Comment by dongjoon-hyun** [comment]
Thank you, @LuciferYang . 

Before touching those parts, we need to make it sure that this RocksDB has no regression. Could you run benchmarks and include them?

**Comment by LuciferYang** [comment]
> Before touching those parts, we need to make it sure that this RocksDB has no regression. Could you run benchmarks and include them?

**Comment by dongjoon-hyun** [comment]
Since this is `RocksDB` update, cc @viirya and @HeartSaVioR

**Comment by viirya** [comment]
> [error] sbt.librarymanagement.ResolveException: Error downloading org.rocksdb:rocksdbjni:6.29.3

Is it released?

**Comment by LuciferYang** [comment]
> Is it released?

From the release notes , it has been released for 3 days. 

I'll go to the Facebook community to find the reason and I'll ping you when this pr ready for review.

Sorry

**Comment by dongjoon-hyun** [comment]
Is there any update, @LuciferYang ?

**Comment by LuciferYang** [comment]
> Is there any update, @LuciferYang ?

Unfortunately, I didn't get more useful information. Shall I close this pr and reopen it after the problem is solved?

**Comment by HyukjinKwon** [comment]
Up to you. If that looks going to be released soon, I don't mind leaving it open.

**Comment by LuciferYang** [comment]
> Up to you. If that looks going to be released soon, I don't mind leaving it open.

OK

**Comment by dongjoon-hyun** [comment]
BTW, Apache Spark 3.3 branch cut is planned on May 15th.

**Comment by LuciferYang** [comment]
> BTW, Apache Spark 3.3 branch cut is planned on May 15th.

@adamretter has not replied to the comments on GitHub for the time being. I just sent an email to him and hope to get a reply. If there is no release plan in the short term, I will close this pr.

**Comment by dongjoon-hyun** [comment]
Thank you for new information. You don't need to close this PR~
I just want to share our due date for Apache Spark 3.3.0.
If we miss 3.3, we can try this at 3.4, @LuciferYang .

**Comment by dongjoon-hyun** [comment]
We got the official confirmation like the following. `6.29.3` fails. We need a new version of rocksDB and rocksDB JNI will arrive after that one or two months.
- https://github.com/facebook/rocksdb/issues/9550#issuecomment-1058515642

**Comment by LuciferYang** [comment]
@dongjoon-hyun Thank you for syncing.

**Comment by adamretter** [comment]
@LuciferYang I think you missed an `@` when you mentioned me - so I did not get any notification. There is a problem at the moment with performing the releases due to a regression that was introduced by someone else trying to introduce Apple Silicon support - I am looking into this but it will take a little time. Details are here - https://github.com/facebook/rocksdb/pull/9254#issuecomment-1054598516

**Comment by LuciferYang** [comment]
@dongjoon-hyun I found @adamretter is fixing the issue of java version release:

- [Fix RocksJava releases for macOS](https://github.com/facebook/rocksdb/pull/9662)

**Comment by dongjoon-hyun** [comment]
For a record, `RocksDB` 7.0.2 is released.
- https://github.com/facebook/rocksdb/releases/tag/v7.0.2

**Comment by LuciferYang** [comment]
> For a record, `RocksDB` 7.0.2 is released.
> 
> * https://github.com/facebook/rocksdb/releases/tag/v7.0.2

Yes, but 7.0.2 also cannot be downloaded through Maven repository, and it seems that @pdillinger and @adamretter wantsto release 6.29.4 first:
- https://github.com/facebook/rocksdb/pull/9678

**Comment by LuciferYang** [comment]
wait https://github.com/facebook/rocksdb/issues/9689

**Comment by adamretter** [comment]
@LuciferYang 6.29.4 is now available

**Comment by dongjoon-hyun** [comment]
Thank you, @adamretter !

**Comment by dongjoon-hyun** [comment]
It seems that 6.29.4.1 exists too. Are we supposed to use `6.29.4.1`, @adamretter ?
- https://repo1.maven.org/maven2/org/rocksdb/rocksdbjni/6.29.4/
- https://repo1.maven.org/maven2/org/rocksdb/rocksdbjni/6.29.4.1/

**Comment by adamretter** [comment]
I just added 6.29.4.1 as the full-fat jar for 6.29.4 was missing the native windows binary. If you don't use the full fat jar or don't use Windows, then it's not an issue. If you do use Windows, you could use the Windows specific jar even then

**Comment by LuciferYang** [comment]
Thank you, @adamretter !

**Comment by LuciferYang** [comment]
Manually restarted GA

**Comment by LuciferYang** [comment]
> I just added 6.29.4.1 as the full-fat jar for 6.29.4 was missing the native windows binary. If you don't use the full fat jar or don't use Windows, then it's not an issue. If you do use Windows, you could use the Windows specific jar even then

Yes, 6.29.4 is missing `librocksdbjni-win64.dll`

**Comment by adamretter** [comment]
@LuciferYang I have also now released RocksJava 7.0.3 to Maven Central

**Comment by LuciferYang** [comment]
Thanks @adamretter !

**Comment by LuciferYang** [comment]
7.0.3
```
[INFO] Running org.apache.spark.util.kvstore.RocksDBBenchmark
                                        	count	mean	min	max	95th
dbClose                                 	4	0.332	0.235	0.557	0.557
dbCreation                              	4	75.354	3.301	290.771	290.771
naturalIndexCreateIterator              	1024	0.005	0.002	1.418	0.007
naturalIndexDescendingCreateIterator    	1024	0.006	0.006	0.068	0.008
naturalIndexDescendingIteration         	1024	0.006	0.004	0.261	0.007
naturalIndexIteration                   	1024	0.006	0.004	0.061	0.009
randomDeleteIndexed                     	1024	0.026	0.020	0.231	0.035
randomDeletesNoIndex                    	1024	0.014	0.012	0.040	0.017
randomUpdatesIndexed                    	1024	0.081	0.032	31.135	0.085
randomUpdatesNoIndex                    	1024	0.037	0.033	0.528	0.042
randomWritesIndexed                     	1024	0.119	0.034	54.771	0.122
randomWritesNoIndex                     	1024	0.042	0.036	1.448	0.048
refIndexCreateIterator                  	1024	0.005	0.004	0.020	0.006
refIndexDescendingCreateIterator        	1024	0.003	0.003	0.029	0.005
refIndexDescendingIteration             	1024	0.006	0.005	0.046	0.008
refIndexIteration                       	1024	0.007	0.005	0.072	0.009
sequentialDeleteIndexed                 	1024	0.021	0.017	0.100	0.026
sequentialDeleteNoIndex                 	1024	0.014	0.012	0.039	0.017
sequentialUpdatesIndexed                	1024	0.047	0.040	0.747	0.058
sequentialUpdatesNoIndex                	1024	0.040	0.029	0.730	0.049
sequentialWritesIndexed                 	1024	0.050	0.044	1.842	0.057
sequentialWritesNoIndex                 	1024	0.041	0.031	2.213	0.044
```

Let me compile with 7.0.3 locally first

**Comment by LuciferYang** [comment]
[ac2f6f3](https://github.com/apache/spark/pull/35580/commits/ac2f6f30047f8372f55d24ae766ccc1e1d56682c) try updating rocksdbjni to 7.0.3

**Comment by LuciferYang** [comment]
6.29.4.1 on Linux

```
[INFO] Running org.apache.spark.util.kvstore.RocksDBBenchmark
                                        	count	mean	min	max	95th
dbClose                                 	4	0.356	0.281	0.519	0.519
dbCreation                              	4	77.093	3.549	300.771	300.771
naturalIndexCreateIterator              	1024	0.006	0.002	1.429	0.008
naturalIndexDescendingCreateIterator    	1024	0.006	0.005	0.071	0.008
naturalIndexDescendingIteration         	1024	0.006	0.004	0.259	0.008
naturalIndexIteration                   	1024	0.006	0.004	0.063	0.009
randomDeleteIndexed                     	1024	0.026	0.019	0.319	0.035
randomDeletesNoIndex                    	1024	0.015	0.012	0.038	0.017
randomUpdatesIndexed                    	1024	0.081	0.032	30.418	0.087
randomUpdatesNoIndex                    	1024	0.063	0.059	0.629	0.068
randomWritesIndexed                     	1024	0.119	0.034	52.506	0.125
randomWritesNoIndex                     	1024	0.071	0.063	1.647	0.078
refIndexCreateIterator                  	1024	0.004	0.004	0.017	0.006
refIndexDescendingCreateIterator        	1024	0.003	0.003	0.031	0.005
refIndexDescendingIteration             	1024	0.007	0.005	0.048	0.009
refIndexIteration                       	1024	0.007	0.005	0.074	0.010
sequentialDeleteIndexed                 	1024	0.021	0.017	0.124	0.024
sequentialDeleteNoIndex                 	1024	0.015	0.012	0.050	0.017
sequentialUpdatesIndexed                	1024	0.070	0.051	1.724	0.082
sequentialUpdatesNoIndex                	1024	0.067	0.050	0.790	0.079
sequentialWritesIndexed                 	1024	0.073	0.058	2.074	0.083
sequentialWritesNoIndex                 	1024	0.057	0.051	3.193	0.062
```

6.29.4.1 on M1

```
[INFO] Running org.apache.spark.util.kvstore.RocksDBBenchmark
                                        	count	mean	min	max	95th
dbClose                                 	4	0.563	0.428	0.861	0.861
dbCreation                              	4	312.788	134.860	737.943	737.943
naturalIndexCreateIterator              	1024	0.003	0.001	0.963	0.003
naturalIndexDescendingCreateIterator    	1024	0.003	0.002	0.048	0.003
naturalIndexDescendingIteration         	1024	0.003	0.003	0.018	0.004
naturalIndexIteration                   	1024	0.004	0.003	0.084	0.011
randomDeleteIndexed                     	1024	0.017	0.013	0.306	0.020
randomDeletesNoIndex                    	1024	0.010	0.009	0.032	0.011
randomUpdatesIndexed                    	1024	0.044	0.020	15.237	0.041
randomUpdatesNoIndex                    	1024	0.019	0.015	0.533	0.022
randomWritesIndexed                     	1024	0.091	0.023	50.713	0.088
randomWritesNoIndex                     	1024	0.036	0.023	7.722	0.035
refIndexCreateIterator                  	1024	0.002	0.002	0.013	0.002
refIndexDescendingCreateIterator        	1024	0.001	0.001	0.010	0.002
refIndexDescendingIteration             	1024	0.010	0.003	6.005	0.005
refIndexIteration                       	1024	0.005	0.004	0.233	0.005
sequentialDeleteIndexed                 	1024	0.015	0.013	0.088	0.018
sequentialDeleteNoIndex                 	1024	0.013	0.011	0.325	0.015
sequentialUpdatesIndexed                	1024	0.023	0.018	0.512	0.029
sequentialUpdatesNoIndex                	1024	0.026	0.021	0.536	0.031
sequentialWritesIndexed                 	1024	0.041	0.023	1.817	0.082
sequentialWritesNoIndex                 	1024	0.030	0.024	2.137	0.039
```

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDB.scala [hunk 1]
```diff
@@ -177,7 +177,7 @@ class RocksDB(
         numKeysOnWritingVersion -= 1
       }
     }
-    writeBatch.remove(key)
+    writeBatch.delete(key)
```
Comment (LuciferYang) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDB.scala:180]: The `remove` series method of `WriteBatch` is identified as `@Deprecated` in 6.x and [deleted in 7.x](https://github.com/facebook/rocksdb/pull/9576), here use `delete` instead 

Am I right @adamretter ？

**Comment by LuciferYang** [comment]
@dongjoon-hyun Both 6.29.4.1 and 7.0.3 can pass GA. I chose 7.x because in addition to supporting M1, it is also a version compiled by Java 8, while 6.x is still compiled by Java 7.

image

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDB.scala (hunk 1 above)
Comment (adamretter) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDB.scala:180]: Yes.

**Comment by adamretter** [comment]
@LuciferYang Just to mention that - most methods that move bytes now have Java `ByteBuffer` versions as well as the older `byte[]` versions. If you are not using the ByteBuffer versions, then you might like to try them, as we have seen improved performance with those.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDB.scala (hunk 1 above)
Comment (LuciferYang) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDB.scala:180]: thanks @adamretter

**Comment by LuciferYang** [comment]
> @LuciferYang Just to mention that - most methods that move bytes now have Java `ByteBuffer` versions as well as the older `byte[]` versions. If you are not using the ByteBuffer versions, then you might like to try them, as we have seen improved performance with those.

Thanks for your advice. I'll try this later, maybe a new pr for Spark

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM. Merged to master for Apache Spark 3.4.
Thank you, @LuciferYang ,  @adamretter and all.

**Comment by LuciferYang** [comment]
thanks all

**Pull request closed without merging**